src/Router.ts

```diff
@@ -0,0 +1,186 @@
+	protected _evented: Evented;
+	protected _isRun: boolean;
+
+	canceled: NavigationArgs;
+	current: NavigationArgs;
+	error: (error: Error) => void;
+	source: RouterSource;
+
+	constructor(kwArgs: RouterArgs) {
+		super(kwArgs);
+
+		this.error = kwArgs.error || null;
+		this.source = kwArgs.source || defaultSource;
+		this._evented = new Evented();
+		this._isRun = false;
+	}
+
+	protected _createEvent(type: string, args: NavigationArgs): CancelableNavigationArgs {
+		const event = <CancelableNavigationArgs> mixin(Object.create(null), args);
+		event.type = type;
+
+		return event;
+	}
+
+	protected _getEventPromise(
+		event: CancelableNavigationArgs,
+		path: string,
+		object: any,
+		method: string
+	): Promise<any> {
+		return new Promise(function (resolve: Function, reject: Function): void {
+			if (typeof object[method] !== 'function') {
+				return resolve();
+			}
+
+			event.preventDefault = function () {
+				reject(new CancelNavigationError('Navigation canceled for path ' + path));
+			};
+
+			const handle = after(object, method, function () {
+				handle.destroy();
+				resolve();
+			});
+			object[method](event);
+		});
+	}
+
+	protected _getRouteArgs(path: string): NavigationArgs {
+		let args: NavigationArgs = null;
+
+		for (let route of this.routes) {
+			args = route.match(path);
+
+			if (args) {
+				break;
+			}
+		}
+
+		return args || this.defaultRoute && this.defaultRoute.match(path);
+	}
+
+	protected _navigate(path: string, args: NavigationArgs): Promise<any> {
+		const current = args.route;
+		const previous = this.current && this.current.route;
+
+		if (current === previous) {
+			return Promise.resolve(current.change && current.change(args));
+		}
+
+		const previousArgs = this.current;
+		const event = this._createEvent('beforechange', args);
+		const getEventPromise = this._getEventPromise.bind(this, event, path);
+
+		const self = this;
+		return getEventPromise(this, 'emit')
+			.then(function () {
+				return previous && getEventPromise(previous, 'beforeExit');
+			})
+			.then(function () {
+				return getEventPromise(current, 'beforeEnter');
+			})
+			.then(function () {
+				return previous && previous.exit && previous.exit();
+			})
+			.then(function () {
+				self.current = args;
+				self.current.route = current;
+
+				return current.enter(path);
+			})
+			.then(function () {
+				self.source && self.source.go(PathRule.join(self.path, path), args.state);
+				self.emit(self._createEvent('change', args));
+			})
+			.catch(function (error: any) {
+				self.canceled = args;
+				self.current = previousArgs;
+
+				if (self.error) {
+					return self.error(error);
+				}
+
+				if (error.name !== 'CancelNavigationError') {
+					throw error;
+				}
+			});
+	}
+
+	destroy(): void {
+		this.go = function () {
+			return Promise.resolve(null);
+		};
+		this.source && this.source.destroy();
+		this.routes = this._rule = this.source = null;
+	}
+
+	emit(data: EventObject): void {
```

Comment:
This method perhaps should be removed, since emitting events is something handled by `Router` itself and not other objects.